### ISSUE TITLE
Exporting paths from compiler packages

### ISSUE THREAD

**Issue by lpw25** [ocaml/opam#3032]
With the namespaces proposal (https://github.com/lpw25/namespaces) I think people will probably want to install things in `<switch>/lib/ocaml/site-packages/foo` for a package `foo`. At the moment people have install files with things like `lib: [ ... ]` which puts things in `<switch>/lib/foo` and there are variables like `lib:foo` that also give you `<switch>/lib/foo`. Ideally I think I would like to make something like: `ocaml:foo` (equal to `<switch>/lib/ocaml/site-packages/foo`) available in the same way. Obviously this set of variables is OCaml specific, so it should really be provided by the `ocaml` package in some way.

Looking at https://opam.ocaml.org/doc/2.0/Manual.html#switchconfig it seems that `lib` and friends come from the `paths` field of the `<switch-prefix>/.opam-switch/switch-config` file. So I think what I am asking for is support for a similar `paths` field in `<pkgname>.install` files, and a corresponding generalization of the variable syntax to allow use of these paths. I guess it would also require similar generalization in `opam-installer`.

I think this should be resolved for opam 2.0 if possible since it will make using the new namespaces proposal much more natural.

**Comment by dbuenzli** [comment]
> Obviously this set of variables is OCaml specific, so it should really be provided by the ocaml package in some way.

It may be a bit difficult to make the `.install` file syntax depend on the package install base. One way out would be to allow `x-field:`  extensions in `.install` files, (i.e. `x-ocaml:` in our case) and provide a mecanism in opam that allow packages to bind paths to the fields of `.install` files.

**Comment by dbuenzli** [comment]
Slightly related to https://github.com/ocaml/opam/issues/3008

**Comment by lpw25** [comment]
Yeah, I was thinking of allowing something like `ocaml:site: ...` in `install` files for this, along with some kind of command-line option to `opam-installer` like `-path ocaml:site=...`.

**Comment by AltGr** [comment]
Some details:
- The switch config file allows to redefine the base directories, _e.g._ `lib`. The packages' lib directories would then become `<libdir>/pkgname`. There is otherwise no customisation available in `opam-installer`, and that might confuse external tools.
- There is no facility for generating it, but the package configuration files can be generated upon package build and define any variable they like (although, if redefining predefined variables, the behaviour is currently unspecified). So if you were to define a `namespace-path` variable in `pkg`'s `.config` file, you could then access it as `<pkg>:namespace-path` from any dependent package. The `ocaml` package could of course be used in that way (it already generates a `.config` file that defines a few variables like `ocaml:native-dynlink` (boolean) or `ocaml:stubsdir` (string).

**Comment by lpw25** [comment]
It also seems like `opam-installer` is a bit ocaml specific, including things like `stublibs` and `toplevel`. So a change along these lines could make it less OCaml specific.

> you could then access it as <pkg>:namespace-path

That doesn't apply to `.install` files right? I also think using `"%{ocaml:site}%/foo" is noticeably less convenient than `ocaml:site:foo`.

It is probably worth noting that something like this would also be useful if you were, for example, trying to use opam to manage python libraries -- since python also expects libraries to be installed in a directory like `<prefix>/lib/python/site-packages/`.
